---
title: "Data Integration Tutorial"
author: "Zain Arifin"
date: "`r format(Sys.Date(), '%B %d, %Y')`"
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: show
    highlight: zenburn
    theme: simplex
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{css, echo=FALSE}
.list-group-item.active, .list-group-item.active:hover, .list-group-item.active:focus {
    color: #ffffff;
    background-color: #333333;
}
p {
  color: #333333;
}
```

## Introduction 

Today, we will have a look at data integration with MOFA2. Briefly, the data consists of four omics including DNA methylation, RNA-seq, somatic mutations and drug response data from blood for N=200 patients with Chronic Lymphocytic Leukemia (CLL). The data set is explained in detail in this article: https://www.jci.org/articles/view/93801. One caveat about MOFA2 is that it can only be used for integrating data that are matched. For unmatched, please have a look at mechanistic-based or graph-based integration.

## Load data

```{r, load_libraries, message=FALSE, warning=FALSE}
options(future.globals.maxSize = 10 * 1024^3) 

library(MOFA2)
library(MOFAdata)
library(data.table)
library(tidyverse)
```

```{r, read_data, cache=F}
utils::data("CLL_data")       
lapply(CLL_data,dim)

CLL_data$Drugs[1:5,1:5]
CLL_data$Methylation[1:5,1:5]
CLL_data$Mutations[1:5,1:5]
CLL_data$mRNA[1:5,1:5]

CLL_metadata <- fread("ftp://ftp.ebi.ac.uk/pub/databases/mofa/cll_vignette/sample_metadata.txt")
head(CLL_metadata)
```

Sample metadata are stored as a data.frame. Important columns are:

Gender: m (male), f (female)
Age: age in years
TTT: time (in years) which passed from taking the sample to the next treatment
TTD: time (in years) which passed from taking the sample to patientsâ€™ death
treatedAfter: (TRUE/FALSE)
Died: whether the patient died (TRUE/FALSE)

## Run MOFA

Prior to running, please make sure that each omic data has been individually normalized. There is an argument to also scale the data, such that each omic should be in the same numerical scale (if data is numeric). However, this is entirely dependent on what data you are working with. For more information on MOFA, have a look at their FAQ page: https://biofam.github.io/MOFA2/faq.html

```{r, run_mofa}
MOFAobject <- create_mofa(CLL_data)
plot_data_overview(MOFAobject)

data_opts <- get_default_data_options(MOFAobject)
data_opts

model_opts <- get_default_model_options(MOFAobject)
model_opts

train_opts <- get_default_training_options(MOFAobject)
train_opts$convergence_mode <- "slow"
train_opts$seed <- 110

train_opts

MOFAobject <- prepare_mofa(MOFAobject,
  data_options = data_opts,
  model_options = model_opts,
  training_options = train_opts
)

MOFAobject <- run_mofa(MOFAobject,
                       outfile="MOFA2_CLL.hdf5", 
                       use_basilisk = T)
# The first training you do with MOFA will take some time as they will install the required Python dependency

# Add sample metadata to the model
samples_metadata(MOFAobject) <- CLL_metadata
```

```{r, check_mofa}
plot_factor_cor(MOFAobject)
plot_variance_explained(MOFAobject, max_r2=15)
plot_variance_explained(MOFAobject, plot_total = T)[[2]]
correlate_factors_with_covariates(MOFAobject, 
  covariates = c("died","age"), 
  plot="log_pval"
)
```


We several interesting output from evaluating the factors. First, we make sure that there is not much auto-correlation between factors. If there is, it is a good idea to retrain the model and reduce the number of factors. Similar to PCA, MOFA2 puts embedings which highlight the greatest source of variation if Factor1 (PC1 in PCA, for instance). We see that factor 1 captures many variation accross all dataset. We also see that factor 1 showed a small correlation to "died condition".


```{r, viz_factor_mofa}
p <- plot_factor(MOFAobject, 
  factors = c(1,2,3),
  color_by = "died",
  dot_size = 3,        # change dot size
  dodge = TRUE,           # dodge points with different colors
  legend = TRUE,          # remove legend
  add_violin = TRUE,      # add violin plots,
  violin_alpha = 0.25  # transparency of violin plots
)
p
```

## Analyze MOFA weights

```{r, weights_mofa}
plot_weights(MOFAobject,
 view = "Mutations",
 factor = 1,
 nfeatures = 20,     # Top number of features to highlight
 scale = T           # Scale weights from -1 to 1
)

plot_weights(MOFAobject,
 view = "mRNA",
 factor = 1,
 nfeatures = 20,     # Top number of features to highlight
 scale = T           # Scale weights from -1 to 1
)

plot_weights(MOFAobject,
 view = "Drugs",
 factor = 1,
 nfeatures = 20,     # Top number of features to highlight
 scale = T           # Scale weights from -1 to 1
)

plot_weights(MOFAobject,
 view = "Methylation",
 factor = 1,
 nfeatures = 20,     # Top number of features to highlight
 scale = T           # Scale weights from -1 to 1
)
```

We see that for mutation IGHV is one of the main source of variation in factor. Lets confirm this with UMAP.

```{r, viz_UMAP_mofa}
set.seed(1)
MOFAobject <- run_umap(MOFAobject)

plot_dimred(MOFAobject,
  method = "UMAP", 
  color_by = "IGHV",
  dot_size = 5
)
plot_dimred(MOFAobject,
  method = "UMAP", 
  color_by = "died",
  dot_size = 5
)

# Check others 

# BRAF mutation
plot_dimred(MOFAobject,
  method = "UMAP",
  color_by = "BRAF",
  dot_size = 5, groups = "all"
)

# ENSG00000168594/ADAM29 expression
plot_dimred(MOFAobject,
  method = "UMAP",
  color_by = "ENSG00000168594",
  dot_size = 5, groups = "all"
)
```


There are more analyses that can be done with MOFA2. Please have a look at MOFA documentations and try to complete the following task:

1. Perform survival analysis using Factor 1 (or other factors) as weight
2. Run gene set enrichment (GSEA) analysis or over-representation analysis on Factor 1
3. Explore Drugs and methylation variables and identify its biological significane towards CLL.
4. From the data integration, what conclusion can you pull when you combine these 4 omics layers? Try to link up all the findings you get from each layer and build up a possible biological explanation.